models/cartDao.js

```diff
@@ -0,0 +1,93 @@
+const appDataSource = require('./dataSource');
+
+const createCart = async (userId, productId, sizeId) => {
+  const insertCart = await appDataSource.query(`
+    INSERT INTO carts 
+    (
+      user_id,
+      product_option_id
+    ) 
+    SELECT ${userId}, (SELECT po.id FROM product_options AS po
+    JOIN products AS p ON po.product_id = p.id
+    JOIN sizes AS s ON po.size_id = s.id
+    WHERE po.product_id=${productId} AND po.size_id=${sizeId}) 
+    WHERE NOT EXISTS
+    (
+      SELECT * FROM carts AS c 
+      WHERE c.user_id=${userId} AND c.product_option_id= 
+        (
+          SELECT po.id FROM product_options AS po
+          JOIN products AS p ON po.product_id = p.id
+          JOIN sizes AS s ON po.size_id = s.id
+          WHERE po.product_id=${productId} AND po.size_id=${sizeId}
+        )  
+    )
+  `)
+  
+  if (insertCart.affectedRows === 0) {
+    const error = new Error('FAILED');
+    error.statusCode = 400;
+    throw error;
+  }
+  return insertCart;
+}
+
+const getCart = async (userId) => {
+  const result = await appDataSource.query(`
+    SELECT
+      c.user_id AS userId,
+      c.product_option_id AS productOptionId,
+      c.count,
+      po.product_id AS productId,
+      po.size_id AS sizeId,
+      p.thumbnail_image_url AS thumbnailUrl,
+      p.name,
+      p.price,
+      s.foot_size AS footSize,
+      po.stock
+    FROM carts AS c
+    JOIN users AS u ON c.user_id = u.id
+    JOIN product_options AS po ON c.product_option_id = po.id
+    JOIN products p ON po.product_id = p.id
+    JOIN sizes AS s ON po.size_id = s.id
+    WHERE c.user_id=?`,
+    [userId]
+  );
+  return result;
+}
+
+const updateCart = async (userId, count, stock) => {
+  const result = await appDataSource.query(`
+    UPDATE carts AS c, product_options AS po
+    SET count=${count}
+    WHERE user_id=${userId} AND stock=${stock} AND ${count} <= ${stock}
+  `);
```

Comment:
장바구니에 넣었을 때 productOptions 테이블의 stock값도 변경을 해야된다면 장바구니에 추가된 count만큼 stock에서 줄여주어야 할 것 같습니다!
where 절에서의 조건도 좀 더 명확한 조건을 한번 고민해주세요!

```suggestion
const updateCart = async (userId, count, stock) => {
  const result = await appDataSource.query(`
    UPDATE carts AS c, product_options AS po
    SET c.count=${count}, po.stock = stock - ${count}
    WHERE user_id=${userId} AND stock=${stock} AND ${count} <= ${stock}
  `);
```

Reply:
음.. 이 부분은 stock의 경우는 장바구니 기능 자체가 여러 user가 장바구니에 담아도 실제 주문으로 이루어지지 않으면 stock이 차감되지 않아야 한다고 생각해서 그렇게 로직을 구성하였습니다.